gadopt/energy_solver.py

```diff
@@ -145,16 +121,70 @@ def __init__(
 
         self.timestepper = timestepper
         self.T_old = Function(self.Q)
+
+        rho_cp = approximation.rhocp()
+        eq_terms = [
+            scal_eq.scalar_advection_term,
+            scal_eq.scalar_diffusion_term,
+            scal_eq.scalar_absorption_term,
+            scal_eq.scalar_source_term,
+        ]
+        terms_kwargs = {
+            "advective_velocity_scaling": rho_cp,
+            "diffusivity": approximation.kappa(),
+            "reference_for_diffusion": approximation.Tbar,
+            "sink": approximation.linearized_energy_sink(u),
+            "source": approximation.energy_source(u),
+            "u": u,
```

Comment:
I like that the equation is fairly generic, but I don't like the discoverability of having to provide the various terms this way. How are they documented? How are they enforced, i.e. in case of a typo?

I know it's not very different to before, but just something that comes to mind when reading this new approach is that everything is stuffed in a fairly opaque "kwargs" bucket.

Reply:
I agree with what you raised here. I tried to provide some logic in the Equation class based on attributes I added to each term function. Let me know what you think of it.

Reply:
Definitely an improvement, yeah. Runtime validation is nice, and being able to see all the required attributes in one place at least makes it a bit easier to figure out documentation and discoverability.